### PR TITLE
Add `no-single-promise-in-promise-methods` rule

### LINKED ISSUE #2256: Rule proposal: `no-single-promise-in-promise-methods`
### Description

Disallow using `Promise` method with a single element array as parameter.

### Fail

```js
Promise.all([promise]);

Promise.any([promise]);

Promise.race([promise]);
```

### Pass

```js
Promise.all([promise, anotherPromise]);
Promise.all(notArrayLiteral);
Promise.all([...promises]);

Promise.any([promise, anotherPromise]);

Promise.race([promise, anotherPromise]);

Promise.allSettled([promise]);
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by Clement398**
Fixes #2256

### CODE REVIEW

**Comment by sindresorhus** [comment]
Same feedback as in https://github.com/sindresorhus/eslint-plugin-unicorn/pull/2259

File: rules/no-single-promise-in-promise-methods.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+const messages = {
+	[MESSAGE_ID]: 'Promise.{{ method }} contains an array of a single element.',
+};
+
+const isPromiseMethodWithSinglePromise = node =>
+	node.callee.type === 'MemberExpression'
+	&& node.callee.object.type === 'Identifier'
+	&& node.callee.object.name === 'Promise'
+	&& node.callee.property.type === 'Identifier'
+	&& ['all', 'allSettled', 'any', 'race'].includes(node.callee.property.name)
```
Comment (sindresorhus) [inline rules/no-single-promise-in-promise-methods.js:12]: This part could be reused with your other rule. Maybe move it to a separate method and move it to https://github.com/sindresorhus/eslint-plugin-unicorn/tree/main/rules/utils

**Comment by fisker** [comment]
Should we add an option to skip `Promise.allSettled()`? 

```js
const [{reason: error, value}] = await Promise.allSettled([promise]);
```

can be useful to avoid `try/catch`.

**Comment by Clement398** [comment]
> Same feedback as in #2259

It's corrected.

File: rules/utils/is-promise-method-with-array.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+const METHODS = new Set(['all', 'allSettled', 'any', 'race']);
+
+const isPromiseMethodWithArray = node =>
+	node.callee.type === 'MemberExpression'
+	&& node.callee.object.type === 'Identifier'
+	&& node.callee.object.name === 'Promise'
+	&& node.callee.property.type === 'Identifier'
+	&& METHODS.has(node.callee.property.name)
+	&& node.arguments.length === 1
+	&& node.arguments[0].type === 'ArrayExpression';
```
Comment (fisker) [inline rules/utils/is-promise-method-with-array.js:11]: Please add 

```js
Promise.all([...promises])
```

to tests.

File: rules/no-single-promise-in-promise-methods.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+};
+
+const isPromiseMethodWithSinglePromise = node =>
+	isPromiseMethodWithArray(node)
+	&& node.arguments[0].elements.length === 1;
+
+const getMethodName = node => node.callee.property.name;
+
+const getFixer = ({sourceCode}, node) => fixer =>
+	fixer.replaceText(node, sourceCode.getText(node.arguments[0].elements[0]));
```
Comment (fisker) [inline rules/no-single-promise-in-promise-methods.js:16]: This will break

```js
await Promise.all([(a, b)])
```

File: rules/no-single-promise-in-promise-methods.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		if (isPromiseMethodWithSinglePromise(node)) {
+			context.report({
+				node,
+				messageId: MESSAGE_ID,
+				data: {
+					method: getMethodName(node),
+				},
+				fix: getFixer(context, node),
+			});
+		}
```
Comment (fisker) [inline rules/no-single-promise-in-promise-methods.js:30]:
```suggestion
		if (!isPromiseMethodWithSinglePromise(node)) {
			return;
		}

		return {
			node,
			messageId: MESSAGE_ID,
			data: {
				method: getMethodName(node),
			},
			fix: getFixer(context, node),
		};
```

File: test/no-single-promise-in-promise-methods.mjs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	messageId: 'no-single-promise-in-promise-methods',
+};
+
+test({
+	valid: [
+		outdent`
+			const func = () => Promise.resolve('promise');
+			const promise = func();
+			await Promise.all([func(), promise]);
+		`,
```
Comment (fisker) [inline test/no-single-promise-in-promise-methods.mjs:16]: This test can be simplified to `Promise.all([promise, anotherPromise])`.

File: test/no-single-promise-in-promise-methods.mjs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+		outdent`
+			const func = () => Promise.resolve('promise');
+			const promise = func();
+			const tab = [func(), promise];
+			await Promise.all(tab);
+		`,
```
Comment (fisker) [inline test/no-single-promise-in-promise-methods.mjs:22]: `Promise.all(notArrayLiteral)`.

File: rules/no-single-promise-in-promise-methods.js (hunk 3 above)
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js]: It's corrected.

File: rules/no-single-promise-in-promise-methods.js (hunk 4 above)
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js]: It's corrected.

File: test/no-single-promise-in-promise-methods.mjs (hunk 6 above)
Comment (Clement398) [reply test/no-single-promise-in-promise-methods.mjs:17]: It's corrected.

**Comment by Clement398** [comment]
> Should we add an option to skip `Promise.allSettled()`?
> 
> ```js
> const [{reason: error, value}] = await Promise.allSettled([promise]);
> ```
> 
> can be useful to avoid `try/catch`.

I've excluded the allSettled method from this rule in order to take this usage into account.

File: rules/utils/is-promise-method-with-array.js (hunk 2 above)
Comment (fisker) [reply rules/utils/is-promise-method-with-array.js:13]: You missed the holes again.

Comment (Clement398) [reply rules/utils/is-promise-method-with-array.js]: It's corrected.

File: test/no-single-promise-in-promise-methods.mjs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+const {test} = getTester(import.meta);
+
+const error = {
+	messageId: 'no-single-promise-in-promise-methods',
+};
+
+test({
+	valid: [
+		outdent`
```
Comment (fisker) [inline test/no-single-promise-in-promise-methods.mjs:12]: Please use single quote.

File: rules/utils/is-promise-method-with-array.js (hunk 2 above)
Comment (fisker) [reply rules/utils/is-promise-method-with-array.js]: I didn't see tests for holes.

File: test/no-single-promise-in-promise-methods.mjs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+				promise
+			`,
+		},
+		{
+			code: outdent`
+				Promise.race([(promise, anotherPromise)])
+			`,
+			errors: [error],
+			output: outdent`
+				(promise, anotherPromise)
```
Comment (fisker) [inline test/no-single-promise-in-promise-methods.mjs:132]: This file will break

```js
foo
Promise.race([(promise, anotherPromise)])
```

**Comment by fisker** [comment]
Another edge case

```
-Promise.all([-1])
```

**Comment by fisker** [comment]
I don't think we can auto fix cases when it's not awaited.

```js
const promise = Promise.all([somethingMaybeNotPromise])
await promise.catch(...)
```

call `somethingMaybeNotPromise.catch` will throw.

**Comment by fisker** [comment]
I think the correct logic is:

If awaited, use current fix(auto).
If not, suggest two fixes,
1, use the value directly
2.wrap  in `Promise.resolve()`

File: rules/utils/is-promise-method-with-array.js (hunk 2 above)
Comment (Clement398) [reply rules/utils/is-promise-method-with-array.js]: I added a test for this case.

File: test/no-single-promise-in-promise-methods.mjs (hunk 7 above)
Comment (Clement398) [reply test/no-single-promise-in-promise-methods.mjs]: I changed that.

File: test/no-single-promise-in-promise-methods.mjs (hunk 8 above)
Comment (Clement398) [reply test/no-single-promise-in-promise-methods.mjs]: I have restricted the rule to CallExpression, Identifier and MemberExpression (await or not) to avoid this kind of case.

**Comment by Clement398** [comment]
> I think the correct logic is:
> 
> If awaited, use current fix(auto). If not, suggest two fixes, 1, use the value directly 2.wrap in `Promise.resolve()`

It is done, thanks for the suggestion.

File: rules/no-single-promise-in-promise-methods.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+const MESSAGE_ID_SUGGESTION_2 = 'no-single-promise-in-promise-methods/suggestion-2';
+const messages = {
+	[MESSAGE_ID_ERROR]: 'Parameter in `Promise.{{method}}` should not be a single element array.',
+	[MESSAGE_ID_SUGGESTION_1]: 'Use the value directly.',
+	[MESSAGE_ID_SUGGESTION_2]: 'Wrap the value in a `Promise.resolve`.',
+};
+const METHODS = ['all', 'any', 'race'];
+
+const isPromiseMethodWithSinglePromise = (node, methods) => {
+	const types = new Set(['CallExpression', 'Identifier', 'MemberExpression']);
```
Comment (fisker) [inline rules/no-single-promise-in-promise-methods.js:15]: We should not limit types, it can be anything.

File: rules/no-single-promise-in-promise-methods.js (hunk 3 above)
Comment (fisker) [reply rules/no-single-promise-in-promise-methods.js:35]: I don't know how you fix it, but it's incorrect. Please try not to force push, I can't see changes.

File: test/no-single-promise-in-promise-methods.mjs (hunk 8 above)
Comment (fisker) [reply test/no-single-promise-in-promise-methods.mjs]: The correct fix should check ASI problem instead, we have `./utils/needs-semicolon.js` can be used.

**Comment by Clement398** [comment]
Thank you @fisker for all these improvements.

File: rules/no-single-promise-in-promise-methods.js (hunk 3 above)
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js]: The fix was to limit the types by excluding SequenceExpression.
Sorry about the push force.

File: test/no-single-promise-in-promise-methods.mjs (hunk 8 above)
Comment (Clement398) [reply test/no-single-promise-in-promise-methods.mjs]: I authorized all types and managed the special cases of SpreadElement and SequenceExpression.
I didn't need to use `./utils/needs-semicolon.js`.

File: rules/no-single-promise-in-promise-methods.js (hunk 9 above)
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js]: I authorized all types and managed the special cases of SpreadElement and SequenceExpression.

File: test/snapshots/no-single-promise-in-promise-methods.mjs.md [hunk 10]
```diff
@@ -152,3 +152,147 @@ Generated by [AVA](https://avajs.dev).
...
+
+    `␊
+      1 | foo␊
+    > 2 | Promise.all([[array][0]])␊
+        | ^^^^^^^^^^^^^^^^^^^^^^^^^ Parameter in \`Promise.all\` should not be a single element array.␊
+    ␊
+    --------------------------------------------------------------------------------␊
+    Suggestion 1/2: Use the value directly.␊
+      1 | foo␊
+      2 | [array][0]␊
```
Comment (fisker) [inline test/snapshots/no-single-promise-in-promise-methods.mjs.md:292]: leading semicolon should be added.

File: test/snapshots/no-single-promise-in-promise-methods.mjs.md [hunk 11]
```diff
@@ -152,3 +152,147 @@ Generated by [AVA](https://avajs.dev).
...
+
+    `␊
+      1 | foo␊
+    > 2 | Promise.all([(0, promise)])␊
+        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ Parameter in \`Promise.all\` should not be a single element array.␊
+    ␊
+    --------------------------------------------------------------------------------␊
+    Suggestion 1/2: Use the value directly.␊
+      1 | foo␊
+      2 | (0, promise)␊
```
Comment (fisker) [inline test/snapshots/no-single-promise-in-promise-methods.mjs.md:241]: semi colon

File: test/snapshots/no-single-promise-in-promise-methods.mjs.md (hunk 10 above)
Comment (Clement398) [reply test/snapshots/no-single-promise-in-promise-methods.mjs.md]: It's corrected.

File: test/snapshots/no-single-promise-in-promise-methods.mjs.md (hunk 11 above)
Comment (Clement398) [reply test/snapshots/no-single-promise-in-promise-methods.mjs.md]: It's corrected.

File: test/snapshots/no-single-promise-in-promise-methods.mjs.md [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+
+    `␊
+      1 | await promise␊
+    `
+
+> Error 1/1
+
+    `␊
+    > 1 | await Promise.all([promise])␊
+        |       ^^^^^^^^^^^^^^^^^^^^^^ Parameter in \`Promise.all()\` should not be a single element array.␊
```
Comment (fisker) [inline test/snapshots/no-single-promise-in-promise-methods.mjs.md:25]: Let's set the error range to `callExpression.arguments[0]`?

File: rules/no-single-promise-in-promise-methods.js [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+	&& node.arguments[0].elements[0].type !== 'SpreadElement';
+
+const getText = ({sourceCode}, node, element, prefix = '', suffix = '') => {
+	const previousToken = sourceCode.getTokenBefore(node);
+	const text = sourceCode.getText(element);
+	const parenthesizedText = element.type === 'SequenceExpression' ? `(${text})` : text;
+	const wrappedText = `${prefix}${parenthesizedText}${suffix}`;
+
+	return needsSemicolon(previousToken, sourceCode, wrappedText) ? `;${wrappedText}` : wrappedText;
+};
```
Comment (fisker) [inline rules/no-single-promise-in-promise-methods.js:35]: There is an existing `getParenthesizedText` utility we can use directly.

File: rules/no-single-promise-in-promise-methods.js [hunk 14]
```diff
@@ -0,0 +1,97 @@
+'use strict';
+const {isMethodCall} = require('./ast/index.js');
+const needsSemicolon = require('./utils/needs-semicolon.js');
+
+const MESSAGE_ID_ERROR = 'no-single-promise-in-promise-methods/error';
+const MESSAGE_ID_SUGGESTION_1 = 'no-single-promise-in-promise-methods/suggestion-1';
+const MESSAGE_ID_SUGGESTION_2 = 'no-single-promise-in-promise-methods/suggestion-2';
+const messages = {
+	[MESSAGE_ID_ERROR]: 'Parameter in `Promise.{{method}}()` should not be a single element array.',
```
Comment (fisker) [inline rules/no-single-promise-in-promise-methods.js:9]: How about, ``Wrapping a single element array with `Promise.{{method}}()` is unnecessary.``

File: rules/no-single-promise-in-promise-methods.js [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+	[MESSAGE_ID_SUGGESTION_1]: 'Use the value directly.',
+	[MESSAGE_ID_SUGGESTION_2]: 'Wrap the value in a `Promise.resolve`.',
```
Comment (fisker) [inline rules/no-single-promise-in-promise-methods.js:11]:
```suggestion
	[MESSAGE_ID_SUGGESTION_2]: 'Wrap the value with `Promise.resolve()`.',
```

File: rules/no-single-promise-in-promise-methods.js [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+const getText = ({sourceCode}, node, element, prefix = '', suffix = '') => {
+	const previousToken = sourceCode.getTokenBefore(node);
+	const text = sourceCode.getText(element);
+	const parenthesizedText = element.type === 'SequenceExpression' ? `(${text})` : text;
+	const wrappedText = `${prefix}${parenthesizedText}${suffix}`;
+
+	return needsSemicolon(previousToken, sourceCode, wrappedText) ? `;${wrappedText}` : wrappedText;
+};
+
+const getAutoFixer = (context, node) => fixer => {
```
Comment (fisker) [inline rules/no-single-promise-in-promise-methods.js:37]: Name it `unwrapValue`?

File: rules/no-single-promise-in-promise-methods.js [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+	const [element] = node.arguments[0].elements;
+	const elementWithoutAwait = element.type === 'AwaitExpression' ? element.argument : element;
+
+	return fixer.replaceText(node, getText(context, node, elementWithoutAwait));
+};
+
+const getSuggestion1Fixer = (context, node) => fixer =>
+	fixer.replaceText(node, getText(context, node, node.arguments[0].elements[0]));
+
+const getSuggestion2Fixer = (context, node) => fixer =>
```
Comment (fisker) [inline rules/no-single-promise-in-promise-methods.js:47]: Name it `wrapWithPromiseResolve`

File: rules/no-single-promise-in-promise-methods.js [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+};
+
+const getAutoFixer = (context, node) => fixer => {
+	const [element] = node.arguments[0].elements;
+	const elementWithoutAwait = element.type === 'AwaitExpression' ? element.argument : element;
+
+	return fixer.replaceText(node, getText(context, node, elementWithoutAwait));
+};
+
+const getSuggestion1Fixer = (context, node) => fixer =>
```
[the remark this answers is not in the document]
Comment (fisker) [reply rules/no-single-promise-in-promise-methods.js:44]: Sorry, it can't be, and this is much more tricky than I thought.

[the remark this answers is not in the document]
Comment (fisker) [reply rules/no-single-promise-in-promise-methods.js:44]: I think the current fix will break

```js
Promise.all([1]).then()
```

```js
// Meaningless but valid cases
new Promise.all([A.B]);
```

File: test/snapshots/no-single-promise-in-promise-methods.mjs.md (hunk 12 above)
Comment (Clement398) [reply test/snapshots/no-single-promise-in-promise-methods.mjs.md]: It's done.

File: rules/no-single-promise-in-promise-methods.js (hunk 14 above)
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js]: It's done.

File: rules/no-single-promise-in-promise-methods.js (hunk 15 above)
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js]: It's done.

File: rules/no-single-promise-in-promise-methods.js (hunk 13 above)
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js:53]: It's done.

File: rules/no-single-promise-in-promise-methods.js (hunk 16 above)
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js]: It's done.

File: rules/no-single-promise-in-promise-methods.js (hunk 17 above)
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js]: It's done.

File: rules/no-single-promise-in-promise-methods.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js]: I fixed the error with the .then().
`new Promise.all([A.B])` is a valid case like new `Promise.all([promise])` and does not seem to cause a problem.

File: docs/rules/no-single-promise-in-promise-methods.md [hunk 19]
```diff
@@ -0,0 +1,34 @@
+# Disallow using `Promise` method with a single element array as parameter
```
Comment (sindresorhus) [inline docs/rules/no-single-promise-in-promise-methods.md:1]:
```suggestion
# Disallow passing single-element arrays to `Promise` methods.
```

Comment (Clement398) [reply docs/rules/no-single-promise-in-promise-methods.md]: It's done.

File: test/no-single-promise-in-promise-methods.mjs [hunk 20]
```diff
@@ -6,7 +6,7 @@ const {test} = getTester(import.meta);
 test.snapshot({
 	valid: [
 		'Promise.all([promise, anotherPromise])',
-		'Promise.all(notArrayExpression)',
+		'Promise.all(arrayVariable)',
```
Comment (fisker) [inline test/no-single-promise-in-promise-methods.mjs:9]: Isn't the old name more clear?

File: rules/no-single-promise-in-promise-methods.js [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+const wrapText = (sourceCode, node, element, text, prefix, suffix) => {
+	if (prefix || suffix) {
+		return `${prefix}${text}${suffix}`;
+	}
+
+	if (node.parent.type === 'MemberExpression'
+	&& !isParenthesized(element, sourceCode)
+	&& shouldAddParenthesesToMemberExpressionObject(element, sourceCode)) {
+		return `(${text})`;
+	}
```
Comment (fisker) [inline rules/no-single-promise-in-promise-methods.js:42]: This seems not enough, I haven't test but I guess it will break

```js
await Promise.all([x ** y])
```

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_precedence#table

File: test/no-single-promise-in-promise-methods.mjs (hunk 20 above)
Comment (sindresorhus) [reply test/no-single-promise-in-promise-methods.mjs:9]: I felt it was unclear. It could alternatively be `notArrayLiteral`.

Comment (Clement398) [reply test/no-single-promise-in-promise-methods.mjs]: It's done.

File: rules/no-single-promise-in-promise-methods.js (hunk 21 above)
Comment (Clement398) [reply rules/no-single-promise-in-promise-methods.js]: It's corrected.

Comment (fisker) [reply rules/no-single-promise-in-promise-methods.js]: Not correct, I'll fix.

**Review by fisker — APPROVED** [review]
I think I got everything covered.

Thanks for your work!

**Comment by Clement398** [comment]
Thank you @fisker for all these improvements.
